File: api/pkg/autoscaling/autoscaling.go
package autoscaling

import (
	"database/sql/driver"
	"encoding/json"
	"errors"

	"github.com/caraml-dev/merlin/pkg/deployment"
	merror "github.com/caraml-dev/merlin/pkg/errors"
)

// MetricsType are supported metrics for autoscaling
type MetricsType string

const (
	// CPUUtilization autoscaling based on cpu utilization (0-100 percent)
	CPUUtilization MetricsType = "cpu_utilization"
	// MemoryUtilization autoscaling based on memory utilization (0-100 percent)
	MemoryUtilization MetricsType = "memory_utilization"
	// Concurrency autoscaling based on number of concurrent request that the model should handle
	Concurrency MetricsType = "concurrency"
	// RPS autoscaling based on throughput (request per seconds)
	RPS MetricsType = "rps"
)

var (
	DefaultRawDeploymentAutoscalingPolicy = &AutoscalingPolicy{
		MetricsType: CPUUtilization,
		TargetValue: 50,
	}

	DefaultServerlessAutoscalingPolicy = &AutoscalingPolicy{
		MetricsType: Concurrency,
		TargetValue: 1,
	}
)

// AutoscalingPolicy specify the metrics type and policy value for autoscaling
type AutoscalingPolicy struct {
	// MetricsType type of metrics to be used
	MetricsType MetricsType `json:"metrics_type"`
	// TargetValue specifies the policy value
	TargetValue float64 `json:"target_value"`
}

func (r AutoscalingPolicy) Value() (driver.Value, error) {
	return json.Marshal(r)
}

func (r *AutoscalingPolicy) Scan(value interface{}) error {
	b, ok := value.([]byte)
	if !ok {
		return errors.New("type assertion to []byte failed")
	}

	return json.Unmarshal(b, &r)
}

// ValidateAutoscalingPolicy check autoscaling policy is valid and supported by the given deployment mode
func ValidateAutoscalingPolicy(target *AutoscalingPolicy, mode deployment.Mode) error {
	// raw deployment only support cpu utilization
	if mode == deployment.RawDeploymentMode && target.MetricsType != CPUUtilization {
		return merror.NewInvalidInputErrorf("raw_deployment doesn't support %v autoscaling metrics", target)
	}

	// boundary check for cpu and memory utilization
	if (target.MetricsType == CPUUtilization || target.MetricsType == MemoryUtilization) &&
		(target.TargetValue <= 0 || target.TargetValue > 100) {
		return merror.NewInvalidInputErrorf("policy %v is outside 0-100 range", target.MetricsType)
	}

	// boundary check for rps and concurrency
	if (target.MetricsType == Concurrency || target.MetricsType == RPS) &&
		target.TargetValue <= 0 {
		return merror.NewInvalidInputErrorf("policy %v is less than or equal to 0", target.MetricsType)
	}

	return nil
}

File: python/sdk/merlin/autoscaling.py
from enum import Enum


class MetricsType(Enum):
    """
    Metrics type to be used for AutoscalingPolicy.

    CONCURRENCY: number of concurrent request handled.
    CPU_UTILIZATION: percentage of CPU utilization.
    MEMORY_UTILIZATION: percentage of Memory utilization.
    RPS: throughput in request per second.
    """
    CONCURRENCY = "concurrency"
    CPU_UTILIZATION = "cpu_utilization"
    MEMORY_UTILIZATION = "memory_utilization"
    RPS = "rps"


class AutoscalingPolicy:
    """
    Autoscaling policy to be used for a deployment.

    """

    def __init__(self, metrics_type: MetricsType, target_value: float):
        self._metrics_type = metrics_type
        self._target_value = target_value

    @property
    def metrics_type(self) -> MetricsType:
        """
        Metrics type to be used for the autoscaling
        :return: MetricsType
        """
        return self._metrics_type

    @property
    def target_value(self) -> float:
        """
        Target metrics value when autoscaling should be performed
        :return: target value
        """
        return self._target_value


RAW_DEPLOYMENT_DEFAULT_AUTOSCALING_POLICY = AutoscalingPolicy(MetricsType.CPU_UTILIZATION, 50)
SERVERLESS_DEFAULT_AUTOSCALING_POLICY = AutoscalingPolicy(MetricsType.CONCURRENCY, 1)
